File: c/p019.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */

static int get_length(const struct ListNode* head) {
    int length = 0;
    for (struct ListNode* c = head; c; c = c->next, ++length);

    return length;
}

struct ListNode* removeNthFromEnd(struct ListNode* head, int n){

    const int before_index = get_length(head) - n - 1;

    if (before_index < 0) {
        if (head) {
            return head->next;
        }
        return NULL;
    }


    // iterate to node before target
    struct ListNode* before_node = head;
    for (int i = 0; i < before_index; before_node = before_node->next, ++i);

    before_node->next = before_node->next->next;

    return head;
}

File: python3/p019.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:

        items = [] # "cheating" but only up to 30 items in list anyways
        c = head
        while c:
            items.append(c)
            c = c.next

        before_index = len(items) - n - 1

        # special case: removing first item from list
        if n == len(items):
            return head.next

        # special case: removing last item from list
        if n == 1:
            items[before_index].next = None
            return head

        # normal case, update before to next item
        items[before_index].next = items[before_index+2]

        return head
